[flang][openmp] Adds Parser and Semantic Support for Interop Construct, and Init and Use Clauses.

swatheesh-mcw opened the pull request:
Adds Parser and Semantic Support for the below construct and clauses:
- Interop Construct
- Init Clause
- Use Clause

Note:
The other clauses supported by Interop Construct such as Destroy, Use, Depend and Device are added already.

swatheesh-mcw committed:
Adds Parser and Semantic Support for the below construct and clauses:

- Interop Construct
- Init Clause
- Use Clause

Note:
The other clauses supported by Interop Construct such as Destroy, Use, Depend and Device are added already.

kparzysz commented:
Please follow the common method for implementing modifiers, for example as in https://github.com/llvm/llvm-project/pull/117786.

1. The two modifiers should be prefixed with Omp, i.e. `OmpInteropPreference` and `OmpInteropType`.  Please put them in the "modifiers" namespace in parse-tree.h (all classes are ordered alphabetically).  When a modifier is a wrapper for ENUM_CLASS, the name of the enum should be `Value`, e.g. `ENUM_CLASS(Value, Something, SomethingElse)`.
2. In `OmpInitClause`, please add a line `MODIFIER_BOILERPLATE(OmpInteropPreference, OmpInteropType);`, and delete the WRAPPER_CLASS definitions.  Please do not create nested wrapper classes for individual objects, i.e. delete the InteropVar. The tuple in the class should then be `std::tuple<MODIFIERS(), OmpObject>`.  These macros will define a nested type "Modifier" in the class which is a variant with the two given members.  The tuple will contain an "optional<list<Modifier>>" in place of the MODIFIERS macro.
3. In openmp-parsers.cpp there is a section with parsers for the individual modifiers, in your case, OmpInteropPreference and OmpInteropType.  Please add the parsers there.  Then, in the section below, where all the "OmpXyzClause::Modifier" are parsed, please add a parser for OmpInitClause::Modifier (it's just an alternative parser for the two modifiers).  Then the parser for the OmpInitClause itself would just be `maybe(nonemptyList(Parser<OmpInitClause::Modifier>{}) / ":"), Parser<OmpObject>{}`.
4. In openmp-modifiers.h please add the two new modifiers to the DECLARE_DESCRIPTOR list (in alphabetical order).  In openmp-modifiers.cpp, please add descriptors for these modifiers (also in alphabetical order).  The definitions of the descriptors should be pretty clear based on the other definitions.  The "name" refers to the name in the spec, i.e. "interop-preference" and "interop-type".  The properties for the modifiers are "OmpUnique" for "interop-preference" and "OmpRequired" for the other one.
5. In check-omp-structure.cpp please call `OmpVerifyModifiers`.  To get the optional list of modifiers use `OmpGetModifiers`, and to get a unique modifier, use `OmpGetUniqueModifier`.  To iterate over repeatable modifiers use `OmpGetRepeatableModifier`.  You can find examples of their use in that source file.

swatheesh-mcw commented:
> Please follow the common method for implementing modifiers, for example as in #117786.
> 
>     1. The two modifiers should be prefixed with Omp, i.e. `OmpInteropPreference` and `OmpInteropType`.  Please put them in the "modifiers" namespace in parse-tree.h (all classes are ordered alphabetically).  When a modifier is a wrapper for ENUM_CLASS, the name of the enum should be `Value`, e.g. `ENUM_CLASS(Value, Something, SomethingElse)`.
> 
>     2. In `OmpInitClause`, please add a line `MODIFIER_BOILERPLATE(OmpInteropPreference, OmpInteropType);`, and delete the WRAPPER_CLASS definitions.  Please do not create nested wrapper classes for individual objects, i.e. delete the InteropVar. The tuple in the class should then be `std::tuple<MODIFIERS(), OmpObject>`.  These macros will define a nested type "Modifier" in the class which is a variant with the two given members.  The tuple will contain an "optional<list>" in place of the MODIFIERS macro.
> 
>     3. In openmp-parsers.cpp there is a section with parsers for the individual modifiers, in your case, OmpInteropPreference and OmpInteropType.  Please add the parsers there.  Then, in the section below, where all the "OmpXyzClause::Modifier" are parsed, please add a parser for OmpInitClause::Modifier (it's just an alternative parser for the two modifiers).  Then the parser for the OmpInitClause itself would just be `maybe(nonemptyList(Parser<OmpInitClause::Modifier>{}) / ":"), Parser<OmpObject>{}`.
> 
>     4. In openmp-modifiers.h please add the two new modifiers to the DECLARE_DESCRIPTOR list (in alphabetical order).  In openmp-modifiers.cpp, please add descriptors for these modifiers (also in alphabetical order).  The definitions of the descriptors should be pretty clear based on the other definitions.  The "name" refers to the name in the spec, i.e. "interop-preference" and "interop-type".  The properties for the modifiers are "OmpUnique" for "interop-preference" and "OmpRequired" for the other one.
> 
>     5. In check-omp-structure.cpp please call `OmpVerifyModifiers`.  To get the optional list of modifiers use `OmpGetModifiers`, and to get a unique modifier, use `OmpGetUniqueModifier`.  To iterate over repeatable modifiers use `OmpGetRepeatableModifier`.  You can find examples of their use in that source file.

Sure @kparzysz . Will make the requested change. Thanks!

clementval commented:
Please add a prefix to your title so it is easy to identify the area it touches

`[flang][openmp]`

swatheesh-mcw commented:
> Please follow the common method for implementing modifiers, for example as in #117786.
> 
>     1. The two modifiers should be prefixed with Omp, i.e. `OmpInteropPreference` and `OmpInteropType`.  Please put them in the "modifiers" namespace in parse-tree.h (all classes are ordered alphabetically).  When a modifier is a wrapper for ENUM_CLASS, the name of the enum should be `Value`, e.g. `ENUM_CLASS(Value, Something, SomethingElse)`.
> 
>     2. In `OmpInitClause`, please add a line `MODIFIER_BOILERPLATE(OmpInteropPreference, OmpInteropType);`, and delete the WRAPPER_CLASS definitions.  Please do not create nested wrapper classes for individual objects, i.e. delete the InteropVar. The tuple in the class should then be `std::tuple<MODIFIERS(), OmpObject>`.  These macros will define a nested type "Modifier" in the class which is a variant with the two given members.  The tuple will contain an "optional<list>" in place of the MODIFIERS macro.
> 
>     3. In openmp-parsers.cpp there is a section with parsers for the individual modifiers, in your case, OmpInteropPreference and OmpInteropType.  Please add the parsers there.  Then, in the section below, where all the "OmpXyzClause::Modifier" are parsed, please add a parser for OmpInitClause::Modifier (it's just an alternative parser for the two modifiers).  Then the parser for the OmpInitClause itself would just be `maybe(nonemptyList(Parser<OmpInitClause::Modifier>{}) / ":"), Parser<OmpObject>{}`.
> 
>     4. In openmp-modifiers.h please add the two new modifiers to the DECLARE_DESCRIPTOR list (in alphabetical order).  In openmp-modifiers.cpp, please add descriptors for these modifiers (also in alphabetical order).  The definitions of the descriptors should be pretty clear based on the other definitions.  The "name" refers to the name in the spec, i.e. "interop-preference" and "interop-type".  The properties for the modifiers are "OmpUnique" for "interop-preference" and "OmpRequired" for the other one.
> 
>     5. In check-omp-structure.cpp please call `OmpVerifyModifiers`.  To get the optional list of modifiers use `OmpGetModifiers`, and to get a unique modifier, use `OmpGetUniqueModifier`.  To iterate over repeatable modifiers use `OmpGetRepeatableModifier`.  You can find examples of their use in that source file.

Hi @kparzysz, I tried making the `OmpInteropPreference` and `OmpInteropType` as individual modifiers in `openmp-parsers.cpp`and when tried adding alternative parser for these two modifiers as `OmpInitClause::Modifier`, I am facing the following issues:

1. Since the modifiers are comma separated, adding the parser support in `OmpInitClause::Modifier` as mentioned below results in tuple argument mismatch error.
`TYPE_PARSER(sourced(construct<OmpInitClause::Modifier>(sourced(
     construct<OmpInitClause::Modifier>(maybe(
         verbatim("PREFER_TYPE"_tok) >>
     parenthesized(nonemptyList(Parser<OmpInteropPreference>{})) / ",")),
     construct<OmpInitClause::Modifier>(nonemptyList(Parser<OmpInteropType>{}))))))`
2. If I use `OR (||)` operator instead of `COMMA (,)` to separate the modifiers, then it doesn't throw any error but the parser is not parsing the init clause properly. 
3. I could see in `openmp-parsers.cpp` file, all the other clause modifiers are separated with `OR (||)` operator. I could see a sample of `MobClause` in which it supports parser for comma-separated individual modifiers but it takes object list as the other argument whereas we have just a single object for our `OmpInitClause`.

Can you suggest or direct me through an example where I can refer to create parser support for `OmpInitClause::Modifier`? 

Thanks in advance!

swatheesh-mcw committed:
Modified the declaration and definition of modifiers and addressed review comments.

swatheesh-mcw commented:
> Please follow the common method for implementing modifiers, for example as in #117786.
> 
>     1. The two modifiers should be prefixed with Omp, i.e. `OmpInteropPreference` and `OmpInteropType`.  Please put them in the "modifiers" namespace in parse-tree.h (all classes are ordered alphabetically).  When a modifier is a wrapper for ENUM_CLASS, the name of the enum should be `Value`, e.g. `ENUM_CLASS(Value, Something, SomethingElse)`.
> 
>     2. In `OmpInitClause`, please add a line `MODIFIER_BOILERPLATE(OmpInteropPreference, OmpInteropType);`, and delete the WRAPPER_CLASS definitions.  Please do not create nested wrapper classes for individual objects, i.e. delete the InteropVar. The tuple in the class should then be `std::tuple<MODIFIERS(), OmpObject>`.  These macros will define a nested type "Modifier" in the class which is a variant with the two given members.  The tuple will contain an "optional<list>" in place of the MODIFIERS macro.
> 
>     3. In openmp-parsers.cpp there is a section with parsers for the individual modifiers, in your case, OmpInteropPreference and OmpInteropType.  Please add the parsers there.  Then, in the section below, where all the "OmpXyzClause::Modifier" are parsed, please add a parser for OmpInitClause::Modifier (it's just an alternative parser for the two modifiers).  Then the parser for the OmpInitClause itself would just be `maybe(nonemptyList(Parser<OmpInitClause::Modifier>{}) / ":"), Parser<OmpObject>{}`.
> 
>     4. In openmp-modifiers.h please add the two new modifiers to the DECLARE_DESCRIPTOR list (in alphabetical order).  In openmp-modifiers.cpp, please add descriptors for these modifiers (also in alphabetical order).  The definitions of the descriptors should be pretty clear based on the other definitions.  The "name" refers to the name in the spec, i.e. "interop-preference" and "interop-type".  The properties for the modifiers are "OmpUnique" for "interop-preference" and "OmpRequired" for the other one.
> 
>     5. In check-omp-structure.cpp please call `OmpVerifyModifiers`.  To get the optional list of modifiers use `OmpGetModifiers`, and to get a unique modifier, use `OmpGetUniqueModifier`.  To iterate over repeatable modifiers use `OmpGetRepeatableModifier`.  You can find examples of their use in that source file.

Hi @kparzysz , I have addressed the review comments on [#9e748a6](https://github.com/llvm/llvm-project/pull/120584/commits/9e748a6cc5c42de4a674d0be7082738fae49b7ca). Please let me know if any changes are required. Thanks!

swatheesh-mcw committed:
Preserved the alphabetical ordering of modifier declarations and definitions.

swatheesh-mcw commented:
> > Please follow the common method for implementing modifiers, for example as in #117786.
> > ```
> > 1. The two modifiers should be prefixed with Omp, i.e. `OmpInteropPreference` and `OmpInteropType`.  Please put them in the "modifiers" namespace in parse-tree.h (all classes are ordered alphabetically).  When a modifier is a wrapper for ENUM_CLASS, the name of the enum should be `Value`, e.g. `ENUM_CLASS(Value, Something, SomethingElse)`.
> > 
> > 2. In `OmpInitClause`, please add a line `MODIFIER_BOILERPLATE(OmpInteropPreference, OmpInteropType);`, and delete the WRAPPER_CLASS definitions.  Please do not create nested wrapper classes for individual objects, i.e. delete the InteropVar. The tuple in the class should then be `std::tuple<MODIFIERS(), OmpObject>`.  These macros will define a nested type "Modifier" in the class which is a variant with the two given members.  The tuple will contain an "optional<list>" in place of the MODIFIERS macro.
> > 
> > 3. In openmp-parsers.cpp there is a section with parsers for the individual modifiers, in your case, OmpInteropPreference and OmpInteropType.  Please add the parsers there.  Then, in the section below, where all the "OmpXyzClause::Modifier" are parsed, please add a parser for OmpInitClause::Modifier (it's just an alternative parser for the two modifiers).  Then the parser for the OmpInitClause itself would just be `maybe(nonemptyList(Parser<OmpInitClause::Modifier>{}) / ":"), Parser<OmpObject>{}`.
> > 
> > 4. In openmp-modifiers.h please add the two new modifiers to the DECLARE_DESCRIPTOR list (in alphabetical order).  In openmp-modifiers.cpp, please add descriptors for these modifiers (also in alphabetical order).  The definitions of the descriptors should be pretty clear based on the other definitions.  The "name" refers to the name in the spec, i.e. "interop-preference" and "interop-type".  The properties for the modifiers are "OmpUnique" for "interop-preference" and "OmpRequired" for the other one.
> > 
> > 5. In check-omp-structure.cpp please call `OmpVerifyModifiers`.  To get the optional list of modifiers use `OmpGetModifiers`, and to get a unique modifier, use `OmpGetUniqueModifier`.  To iterate over repeatable modifiers use `OmpGetRepeatableModifier`.  You can find examples of their use in that source file.
> > ```
> 
> Hi @kparzysz , I have addressed the review comments on [#9e748a6](https://github.com/llvm/llvm-project/pull/120584/commits/9e748a6cc5c42de4a674d0be7082738fae49b7ca). Please let me know if any changes are required. Thanks!

I have added another commit [#0fdb9c8](https://github.com/llvm/llvm-project/pull/120584/commits/0fdb9c8e328f717b3393a29076481733fbc003e7) to preserve the alphabetical ordering of modifier declaration and definitions. Please let me know if any modifications are required.

swatheesh-mcw commented:
Ping for review

kiranchandramohan reviewed (commented):
@swatheesh-mcw Thanks for this patch. Could you rebase?

For error messages, we usually have the OpenMP constructs, clauses or reserved terms in caps. Could you update the error messages and tests appropriately?

Could you also add a Lowering test that checks for the TODO message. eg: https://github.com/llvm/llvm-project/blob/main/flang/test/Lower/OpenMP/Todo/dispatch.f90

swatheesh-mcw commented:
Sure @kiranchandramohan, will rebase and update the error messages and tests. Thanks!

swatheesh-mcw committed:
Updated the testfile, rebased and added a test on Todo Lower side for interop construct.

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -5578,6 +5578,98 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+  bool isDependClauseOccured{false};
+  int targetCount{0}, targetSyncCount{0};
+  const auto &dir{std::get<parser::Verbatim>(x.t)};
+  std::list<std::string> ObjectNameList;
```
Don't use strings, use `const Symbol *` to identify objects.  Also, use `std::set` instead of `list`, and check the result of the `insert` operation to see if the element was already in the set.

swatheesh-mcw replied:
Addressed the review comment in [#2370fb7](https://github.com/llvm/llvm-project/pull/120584/commits/2370fb7703a40f53b748168ae3e984cf4d39460e).

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -5578,6 +5578,98 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+  for (const auto &clause : clauseList.v) {
+    common::visit(
+        common::visitors{
+            [&](const parser::OmpClause::Init &InitClause) {
```
Variable names start with a lowercase letter.

swatheesh-mcw replied:
Addressed the review comment in [#2370fb7](https://github.com/llvm/llvm-project/pull/120584/commits/2370fb7703a40f53b748168ae3e984cf4d39460e).

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -5578,6 +5578,98 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+                  }
+                }
+              }
+              const auto &InteropVar{parser::Unwrap<parser::OmpObject>(
```
Lowercase: interopVar.

swatheesh-mcw replied:
Addressed the review comment in [#2370fb7](https://github.com/llvm/llvm-project/pull/120584/commits/2370fb7703a40f53b748168ae3e984cf4d39460e).

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -5578,6 +5578,98 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+                ObjectNameList.push_back(ObjectName);
+              }
+            },
+            [&](const parser::OmpClause::Depend &DependClause) {
```
Lowercase.  There are more places with this issue.

kparzysz reviewed (commented):
Thanks for the changes.

Please avoid using strings to identify anything.  There is almost always a better way to do this, specifically use the corresponding symbols to compare variables (you may need to get the ultimate symbol for that, see `GetUltimate`).

swatheesh-mcw committed:
Addressed review comments and changed object name check with object symbol check.

swatheesh-mcw commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -5578,6 +5578,98 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+                        modifiers)};
+                for (auto it{interopTypeModifier.begin()},
+                     end{interopTypeModifier.end()};
+                     it != end; ++it) {
```
Addressed the review comment in [#2370fb7](https://github.com/llvm/llvm-project/pull/120584/commits/2370fb7703a40f53b748168ae3e984cf4d39460e).

swatheesh-mcw commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -5578,6 +5578,98 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+              const auto ObjectName{name->ToString()};
+              if (ObjectNameList.end() !=
+                  std::find(ObjectNameList.begin(), ObjectNameList.end(),
+                      ObjectName)) {
```
Addressed the review comment in [#2370fb7](https://github.com/llvm/llvm-project/pull/120584/commits/2370fb7703a40f53b748168ae3e984cf4d39460e).

swatheesh-mcw commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -5578,6 +5578,98 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+                     it != end; ++it) {
+                  if (parser::ToUpperCaseLetters(
+                          parser::OmpInteropType::EnumToString((*it)->v)) ==
+                      "TARGETSYNC") {
```
Addressed the review comment in [#2370fb7](https://github.com/llvm/llvm-project/pull/120584/commits/2370fb7703a40f53b748168ae3e984cf4d39460e).

swatheesh-mcw commented on flang/lib/Parser/unparse.cpp:
```diff
@@ -2163,6 +2163,30 @@ class UnparseVisitor {
+  void Unparse(const OmpInitClause &x) {
+    using Modifier = OmpInitClause::Modifier;
+    auto &modifiers{std::get<std::optional<std::list<Modifier>>>(x.t)};
+    bool is_type_start = true;
```
Addressed the review comment in [#2370fb7](https://github.com/llvm/llvm-project/pull/120584/commits/2370fb7703a40f53b748168ae3e984cf4d39460e).

swatheesh-mcw commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -5578,6 +5578,98 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+                  if (targetCount > 1 || targetSyncCount > 1) {
+                    context_.Say(GetContext().directiveSource,
+                        "Each interop-type may be speciﬁed at most once."_err_en_US);
+                  }
```
Addressed the review comment in [#2370fb7](https://github.com/llvm/llvm-project/pull/120584/commits/2370fb7703a40f53b748168ae3e984cf4d39460e).

kparzysz reviewed (approved):
LGTM, thanks.

kiranchandramohan reviewed (commented):
A few nit comments.

swatheesh-mcw committed:
Addressed NIT comments.

swatheesh-mcw commented on flang/lib/Parser/unparse.cpp:
```diff
@@ -2163,6 +2163,30 @@ class UnparseVisitor {
+  void Unparse(const OmpInitClause &x) {
+    using Modifier = OmpInitClause::Modifier;
+    auto &modifiers{std::get<std::optional<std::list<Modifier>>>(x.t)};
+    bool isTypeStart = true;
```
Addressed this in [#ebe9848](https://github.com/llvm/llvm-project/pull/120584/commits/ebe9848a7d1ec65c2d64a43a61f113a61c15bc13).

swatheesh-mcw commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -5578,6 +5578,92 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+              if (llvm::is_contained(objectSymbolList, objectSymbol)) {
+                context_.Say(
+                    GetContext().directiveSource,
+                    "Each interop-var may be speciﬁed for at most one action-clause of each interop construct."_err_en_US);
```
Addressed this in [#ebe9848](https://github.com/llvm/llvm-project/pull/120584/commits/ebe9848a7d1ec65c2d64a43a61f113a61c15bc13).

swatheesh-mcw commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -5578,6 +5578,92 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+  if (isDependClauseOccured && !targetSyncCount) {
+    context_.Say(
+        GetContext().directiveSource,
+        "A depend clause can only appear on the directive if the interop-type includes targetsync"_err_en_US);
```
Addressed this in [#ebe9848](https://github.com/llvm/llvm-project/pull/120584/commits/ebe9848a7d1ec65c2d64a43a61f113a61c15bc13).

swatheesh-mcw commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -5578,6 +5578,92 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+              if (llvm::is_contained(objectSymbolList, objectSymbol)) {
+                context_.Say(
+                    GetContext().directiveSource,
+                    "Each interop-var may be speciﬁed for at most one action-clause of each interop construct."_err_en_US);
```
Addressed this in [#ebe9848](https://github.com/llvm/llvm-project/pull/120584/commits/ebe9848a7d1ec65c2d64a43a61f113a61c15bc13).

swatheesh-mcw commented on flang/lib/Parser/openmp-parsers.cpp:
```diff
@@ -1165,6 +1190,10 @@ TYPE_PARSER(sourced(construct<OmpSimpleStandaloneDirective>(first(
 TYPE_PARSER(sourced(construct<OpenMPSimpleStandaloneConstruct>(
     Parser<OmpSimpleStandaloneDirective>{}, Parser<OmpClauseList>{})))
 
+// 14.1 Interop construct
```
Addressed this in [#ebe9848](https://github.com/llvm/llvm-project/pull/120584/commits/ebe9848a7d1ec65c2d64a43a61f113a61c15bc13).

kiranchandramohan commented:
Could you rebase to address the conflicts in `flang/lib/Parser/unparse.cpp`?

swatheesh-mcw committed:
Changed the InteropRuntimeIdentifier comment on parse-tree.h

swatheesh-mcw commented:
Hi @kiranchandramohan, I have rebased and updated a comment in parse-tree.h for OmpInteropRuntimeIdentifier and pushed the commit. Please let me know if anything else is needed with the patch. Thanks!

kiranchandramohan commented:
Please fix the clang-format issue.

Let me know if you want me to submit this after the format fix.

swatheesh-mcw committed:
Formatted check-omp-structures.cpp

swatheesh-mcw committed:
Formatted using git clang-format

kparzysz commented on flang/test/Lower/OpenMP/Todo/inteorp-construct.f90:
```diff
@@ -0,0 +1,8 @@
+! RUN: %not_todo_cmd %flang_fc1 -emit-fir -fopenmp -fopenmp-version=52 -o - %s 2>&1 | FileCheck %s
```
Please add `!REQUIRES: openmp_runtime` at the beginning of the file.

swatheesh-mcw replied:
Addressed the comment in [#40b5b8d](https://github.com/llvm/llvm-project/pull/120584/commits/40b5b8d499d9d4c9a12f9519dc48be54048537ec).

kparzysz commented on flang/test/Parser/OpenMP/interop-construct.f90:
```diff
@@ -0,0 +1,203 @@
+! RUN: %flang_fc1 -fdebug-unparse -fopenmp-version=52 -fopenmp %s | FileCheck --ignore-case --check-prefix="UNPARSE" %s 
```
Please add `!REQUIRES: openmp_runtime` at the beginning of the file.

swatheesh-mcw committed:
Modified the testcases for Interop construct

kiranchandramohan reviewed (approved):
LGTM. Thanks @swatheesh-mcw for the patch and changes.

swatheesh-mcw commented:
Hi @kiranchandramohan, I could see that the LLVM builds are failing due to failure in written testcases. Is there something I need to do?

Thanks.

kiranchandramohan commented:
@swatheesh-mcw I had to revert the patch in (https://github.com/llvm/llvm-project/pull/132005) because of the failure. I think you have to add `%openmp_flags` to the RUN line of the tests that has `REQUIRES: openmp_runtime`. 

Can you revert https://github.com/llvm/llvm-project/pull/132005 and make the change suggested above?

swatheesh-mcw commented:
Sure @kiranchandramohan, I have created a new pull request [#132343](https://github.com/llvm/llvm-project/pull/132343) to revert [#132005](https://github.com/llvm/llvm-project/pull/132005). I will update the lower testcase there. Thanks!
